From b706651742f94df2909a0f4b399aa3c3f8a6a443 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Fri, 3 Feb 2023 13:40:12 +0100
Subject: [PATCH 021/120] mtd: parsing partitions defined in OF when
 mtdparts/mtdids are empty

When mtdparts and mtdids are empty, the parsing of partitions defined in
device-trees via the `partitions` node with `fixed-partitions` compatible
should be executed one time but it not the case now because
old_mtdparts = mtdparts = old_mtdids = mtdids = NULL
so the first test is always true.

The test on mtd_dev_list_updated() resutl should be done unconditionally
to parse again the device tree is the MTD DEV list change after the call of
mtd_probe_uclass_mtd_devs() / mtd_probe_uclass_spi_nor_devs().

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git a2f378f02b3af54283c24a2716ca3abcc0dac35e mtd: parsing partitions defined in OF when mtdparts/mtdids are empty]
Fixes: dc339bf784f0 ("mtd: add support for parsing partitions defined in OF")
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: Ib7980b3efd19fd542a4b6e60d47f8136fdfd1454
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/289323
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
Domain-Review: Patrice CHOTARD <patrice.chotard@foss.st.com>
---
 drivers/mtd/mtd_uboot.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/drivers/mtd/mtd_uboot.c b/drivers/mtd/mtd_uboot.c
index 69cb3b51f92..330ee09f1f3 100644
--- a/drivers/mtd/mtd_uboot.c
+++ b/drivers/mtd/mtd_uboot.c
@@ -320,16 +320,18 @@ int mtd_probe_devices(void)
 	mtd_probe_uclass_spi_nor_devs();
 
 	/*
-	 * Check if mtdparts/mtdids changed, if the MTD dev list was updated
+	 * Check if the MTD dev list is updated or
+	 * if mtdparts/mtdids changed,
 	 * or if our previous attempt to delete existing partititions failed.
 	 * In any of these cases we want to update the partitions, otherwise,
 	 * everything is up-to-date and we can return 0 directly.
 	 */
-	if ((!mtdparts && !old_mtdparts && !mtdids && !old_mtdids) ||
-	    (mtdparts && old_mtdparts && mtdids && old_mtdids &&
-	     !mtd_dev_list_updated() && !mtd_del_all_parts_failed &&
-	     !strcmp(mtdparts, old_mtdparts) &&
-	     !strcmp(mtdids, old_mtdids)))
+	if (!mtd_dev_list_updated() &&
+	    ((!mtdparts && !old_mtdparts && !mtdids && !old_mtdids) ||
+	     (mtdparts && old_mtdparts && mtdids && old_mtdids &&
+	      !mtd_del_all_parts_failed &&
+	      !strcmp(mtdparts, old_mtdparts) &&
+	      !strcmp(mtdids, old_mtdids))))
 		return 0;
 
 	/* Update the local copy of mtdparts */
-- 
2.47.2

